src/logger.jl

```diff
@@ -0,0 +1,56 @@
+import Logging, Dates
+export ModelLogger, shouldlog, min_enabled_level, catch_exceptions, handle_message
+
+# ------------------------------------------------------------------------------------
+# ModelLogger
+_model_logger_docs = """
+
+    ModelLogger(stream::IO, level::LogLevel)
+
+Based on Logging.SimpleLogger it tries to log all messages in the following format
+
+    [dd/mm/yyyy HH:MM:SS] module source_file:line_number: message
+
+The logger will handle any message from @debug up.
+"""
+struct ModelLogger <: Logging.AbstractLogger
+    stream::IO
+    min_level::Logging.LogLevel
+    message_limits::Dict{Any,Int}
+end
+ModelLogger(stream::IO=stderr, level=Diagnostic) = ModelLogger(stream, level, Dict{Any,Int}())
+
+Logging.shouldlog(logger::ModelLogger, level, _module, group, id) = get(logger.message_limits, id, 1) > 0
+
+Logging.min_enabled_level(logger::ModelLogger) = logger.min_level
+
+Logging.catch_exceptions(logger::ModelLogger) = false
+
+function Logging.handle_message(logger::ModelLogger, level, message, _module, group, id, filepath, line; maxlog = nothing, kwargs...)
+    if maxlog !== nothing && maxlog isa Integer
+        remaining = get!(logger.message_limits, id, maxlog)
+        logger.message_limits[id] = remaining - 1
+        remaining > 0 || return
+    end
+    buf = IOBuffer()
+    iob = IOContext(buf, logger.stream)
+    level = level == Logging.Warn ? "Warning" : string(level)
+    module_name = something(_module, "nothing")
+    file_name = something(filepath, "nothing")
+    line_number = something(line, "nothing")
+    msg_timestamp = Dates.format(Dates.now(), "[dd/mm/yyyy HH:MM:SS]")
+    formatted_message = "$msg_timestamp $module_name $file_name:$line_number: $message"
+    println(iob, formatted_message)
+    write(logger.stream, take!(buf))
+    nothing
+end
+
+# -------------------------------------------------------------------------------------
+# Custom LogStates
+const Diagnostic = Logging.LogLevel(-500)  # Sits between Debug and Info
+const Setup = Logging.LogLevel(500)
```

Comment:
Custom `LogLevel`s not `LogState`s!

It's interesting that you've put these at different importance levels; that would certainly be consistent with them being regarded as custom log levels :+1:

Reply:
Whoops! I'll clean that up.

Reply:
This looks perfect. I think we will have to think a bit about exactly what custom log levels we need. I think both `Setup` (for logging the model setup + construction of key model types?) and `Diagnostic` are excellent. We may also want something like `Simulation`: `Diagnostic` would print the results of diagnostics calculations; `Simulation` would print more general information about the progress of a simulation even when no diagnostics are being calculated.

On the spectrum from `Debug` to `Info` we could have something like

```julia
Debug < Diagnostic < Simulation < Setup < Info
```

This of course doesn't have to be included in the PR here, but something to think about for the future.

Reply:
I'm happy to add those additional levels and structure the levels in that order so we can have a baseline.